security/src/main/java/io/micronaut/security/filters/SecurityFilter.java

```diff
@@ -149,32 +165,76 @@ public int getOrder() {
         String method = request.getMethod().toString();
         String path = request.getPath();
         Map<String, Object> attributes = authentication != null ? authentication.getAttributes() : null;
-        for (SecurityRule rule : securityRules) {
-            SecurityRuleResult result = rule.check(request, routeMatch, attributes);
-            if (result == SecurityRuleResult.REJECTED) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("Unauthorized request {} {}. The rule provider {} rejected the request.", method, path, rule.getClass().getName());
-                }
-                request.setAttribute(REJECTION, forbidden ? HttpStatus.FORBIDDEN : HttpStatus.UNAUTHORIZED);
-                return Publishers.just(new AuthorizationException(authentication));
-            }
-            if (result == SecurityRuleResult.ALLOWED) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("Authorized request {} {}. The rule provider {} authorized the request.", method, path, rule.getClass().getName());
-                }
-                return chain.proceed(request);
-            }
-        }
 
+        return Flowable.fromIterable(orderedSecurityRules)
+                .concatMap(
+                        ordered -> {
+                            if (ordered instanceof SecurityRule) {
+                                SecurityRuleResult result =
+                                        ((SecurityRule) ordered).check(request, routeMatch, attributes);
+                                if (result == SecurityRuleResult.REJECTED || result == SecurityRuleResult.ALLOWED) {
+                                    logResult(result, method, path, ordered);
+                                    return Publishers.just(result);
+                                }
+                                return Publishers.empty();
+                            } else if (ordered instanceof ReactiveSecurityRule) {
+                                return Flowable.fromPublisher(
```

Comment:
You should enforce only a single result by using Single here. As this is its possible to return multiple results which doesn't make sense

Reply:
I tried making this work with Single but for the life of me couldn't since it explodes on the first empty (representing unknown). I added a `firstElement()` call that should cancel after the first. Open to suggestions if you know of a way to make Single work, my reactive is kind of weak.

Reply:
@Steven-M-Brown If you want to support empty then you can use Maybe

Reply:
Having trouble getting a `Maybe` from a `Publisher` short of just using `firstElement()`, suggestions?